api/server/runner_fninvoke.go

```diff
@@ -66,7 +67,19 @@ func (s *Server) handleFnInvokeCall2(c *gin.Context) error {
 		return err
 	}
 
-	return s.ServeFnInvoke(c, app, fn)
+	err = s.ServeFnInvoke(c, app, fn)
+	if _, ok := err.(models.APIError); ok || err == nil {
+		// report all user-directed errors and function responses from here, after submit has run.
+		// this is our never ending attempt to distinguish user and platform errors.
+		ctx, err := tag.New(c.Request.Context(),
+			tag.Upsert(whodunitKey, "user"),
+		)
+		if err != nil {
+			panic(err)
+		}
+		c.Request = c.Request.WithContext(ctx)
+	}
+	return err
```

Comment:
I think we need to move these insertion code to runner. And apply it to cases where we know it's user error. Here, runner could respond with APIError which may or may not be user errors.

Reply:
i'd prefer to do error handling at the highest point possible and feed things into it, especially since we're modifying the context in this specific case in order to do this already. if we need a 'UserAPIError' interface to do that, then so be it (it can implement the `APIError` interface, as well, just with additional method/separate factory method). the thinking in this patch is, we'll start with all APIErrors as being user errors, and any errors we're not expecting (grpc, weird docker error, eg; anything that would 503 normally) are 'service errors'. from that point, we get good coverage (a lot better than now), and can work to blacklist whatever might be 'service' errors; at present, there's no clear definition or list of these. so, we can add facilities to do `UserAPIError` but this will make this patch additionally change probably all errors in `models/` package to `UserAPIError` with maybe a few exceptions (I am not familiar enough with all of our errors to say, but no cases come to mind really) -- up to discuss, intuitively it seems like a bit of a wash without us clearly defining what the difference is.

Reply:
ok. I'm going to propose something relatively similar but for the agent package specifically. here's what I'm thinking:

```go
// in agent package (for use by agent, pure_runner, lb_agent, and transmitted over grpc)
// if this should all be exported to be used from other extenders of fn, we can do that (tell me),
// i didn't think too much about it but it would make sense

// funcError is an error that is the function's fault, that uses the models.APIError but distinguishes fault to function specific errors
type funcError interface {
  models.APIError

  // hidden method needed for duck typing
  userError()
}

type concFuncError models.APIError // maybe embedded, I didn't compile...

func (c concFuncError) userError() { }

func newFuncError(err models.APIError) error { return concFuncError(err) }

func isFuncError(err error) bool { _, ok := err.(funcError); return ok }

// add field 'userError bool' to CallFinished grpc which pure_runner can use isFuncError(err)
// to set the userError field bit in order to distinguish user and service errors over grpc. 
// lb_agent can receive these and recreate the function error (or just normal apierror if not) by 
// doing newFuncError(models.NewAPIError(callproto.status, callproto.msg)) which can then be
// bubbled up to the front end to be consumed as shown in this patch already, which we need
// to do in order to add this to higher level metrics used in middleware (the relevant status code
// metrics)
```